api/src/integrations/core-documents/datasource.js

```diff
@@ -269,12 +271,51 @@ class CoreDocumentsDataSource {
       const transformedAmendment = permitUtils.transformRecord(permit);
       const result = await permitUtils.updateRecord(transformedAmendment, permit);
 
-      if(result.length && result[0].status && result[0].status === 'failure')
+      if (result.length && result[0].status && result[0].status === 'failure')
         throw Error(`permitUtils.updateRecord failed: ${result[0].errorMessage}`);
     } catch (error) {
       throw new Error(`updateAmendment - unexpected error: ${error.message}`);
     }
   }
+
+  /**
+   * Sets the CORE API token and marks when the token will expire
+   *
+   * @memberof CoreDocumentsDataSource
+   */
+  async setClientToken() {
+    console.log('Updating Client Token...');
+    const apiAccess = await getCoreAccessToken(CORE_CLIENT_ID, CORE_CLIENT_SECRET, CORE_GRANT_TYPE);
+    this.apiAccessExpiry = this.getExpiryTime(apiAccess.expires_in);
+    this.client_token = apiAccess.access_token;
+    console.log('Client Token updated.');
+  }
+
+  /**
+   * Gives a time for when the given duration will pass with a buffer
+   *
+   * @param {int} tokenDuration the number of seconds that the token is valid for.
+   * @returns {int} the epoch time when the token is expected to expire ( - the buffer ) = current time + token duration - buffer
+   *
+   * @memberof CoreDocumentsDataSource
+   */
+  getExpiryTime(tokenDuration) {
+    const TIME_BUFFER = 30000;
+    const SECONDS_TO_MILLISECONDS_MULTIPLIER = 1000;
+    return Date.now() + ( tokenDuration * SECONDS_TO_MILLISECONDS_MULTIPLIER ) - TIME_BUFFER;
+  }
+
+  /**
+   * checks if the given time has passed
+   *
+   * @param {int} expiryTime the epoch time we are checking for
+   * @returns {boolean} true if the provided time is in the past
+   *
+   * @memberof CoreDocumentsDataSource
+   */
+  isAPITokenExpired(expiryTime) {
+    return Date.now() >= expiryTime;
+  }
```

Comment:
As an edge case, would you want to artificially increase `Date.now()` to account for time for the API request to be made?

Reply:
that's the TIME_BUFFER. we request a refresh 30 seconds before we actually need it to allow for any delays